ops/testing.py

```diff
@@ -401,9 +402,84 @@ def add_storage(self, storage_name: str, count: int = 1) -> int:
             count: Number of disks being added
 
         Return:
-            The storage_id created
+            A list of storage IDs, e.g. ["my-storage/1", "my-storage/2"].
         """
-        return self._backend.storage_add(storage_name, count)
+        if storage_name not in self._meta.storages:
+            raise RuntimeError(
+                "the key '{}' is not specified as a storage key in metadata".format(storage_name))
+        storage_indices = self._backend.storage_add(storage_name, count)
+
+        # Reset associated cached value in the storage mappings.  If we don't do this,
+        # Model._storages won't return Storage objects for subsequently-added storage.
+        self._model._storages._storage_map[storage_name] = None
+
+        if self.charm is not None and self._hooks_enabled:
```

Comment:
Should we not be checking if `self.charm` is `None` a little earlier in the method and raise an exception if it is?